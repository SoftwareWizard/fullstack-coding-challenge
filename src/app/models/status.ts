export interface Status {
  node_name: string;
}
