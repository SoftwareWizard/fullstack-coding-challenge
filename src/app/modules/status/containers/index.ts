export * from './block-list/block-list.component';
export * from './node-list/node-list.component';
