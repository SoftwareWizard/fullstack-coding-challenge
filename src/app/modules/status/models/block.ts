import { BlockAttributes } from './block-attributes';

export interface Block {
  id: string;
  type: string;
  attributes: BlockAttributes;
}
