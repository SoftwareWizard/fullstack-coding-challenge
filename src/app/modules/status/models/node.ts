export interface Node {
  url: string;
  name: string;
  status?: string;
  isLoading: boolean;
  isOnline: boolean;
  isExpanded: boolean;
}
