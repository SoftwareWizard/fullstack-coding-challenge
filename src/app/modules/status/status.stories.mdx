import { Meta } from '@storybook/addon-docs/blocks';

<Meta title="Status/Introduction" />

# Status Module

The Status Module is responsible for displaying the Status of the Nodes.
It consists of the following Dumb Components:

- Node Detail Component
- Block Detail Component
- Status Component
