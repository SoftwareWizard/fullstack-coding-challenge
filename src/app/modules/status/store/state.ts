import { Node } from '../models';

export class State {
  list: Node[] = [];
}
