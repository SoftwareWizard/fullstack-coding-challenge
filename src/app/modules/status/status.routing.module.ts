import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';

import { NodeListComponent } from './containers';

const routes: Routes = [{ path: '', component: NodeListComponent }];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class StatusRoutingModule {}
