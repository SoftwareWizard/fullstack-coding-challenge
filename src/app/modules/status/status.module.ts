import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';

import { SharedModule } from '../../shared/shared.module';
import { BlockDetailComponent, NodeDetailComponent, NodeStatusComponent } from './components';
import { CounterComponent } from './components/counter/counter.component';
import { BlockListComponent, NodeListComponent } from './containers';
import { StatusRoutingModule } from './status.routing.module';
import { StatusEffects } from './store/status.effects';
import * as statusFacade from './store/status.facade';

const COMPONENTS = [
  NodeListComponent,
  NodeDetailComponent,
  BlockListComponent,
  BlockDetailComponent,
  NodeStatusComponent
];

@NgModule({
  declarations: [...COMPONENTS, CounterComponent],
  imports: [
    CommonModule,
    SharedModule,
    StatusRoutingModule,
    StoreModule.forFeature(statusFacade.featureKey, statusFacade.statusReducer, {
      metaReducers: statusFacade.metaReducers
    }),
    EffectsModule.forFeature([StatusEffects])
  ]
})
export class StatusModule {}
