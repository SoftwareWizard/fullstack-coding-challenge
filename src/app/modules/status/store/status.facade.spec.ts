import { noopAction } from '../../../../../test';
import { Status } from '../models';
import { StatusFacade, statusReducer } from './status.facade';
import { StatusState } from './status.state';

const TEST_NODE_ID = 2;
const TEST_URL = 'http://testurl.com';
const TEST_STATUS = { node_name: 'testNodeName' } as Status;

const TEST_NODE = {
  url: TEST_URL,
  isOnline: undefined,
  isLoading: false,
  isExpanded: false,
  status: undefined
};

const TEST_INITIAL_STATE = {
  nodes: [
    {
      id: 1,
      name: `Node 1`,
      ...TEST_NODE
    },
    {
      id: 2,
      name: `Node 2`,
      ...TEST_NODE
    }
  ]
} as StatusState;

describe('Reducer', () => {
  const facade: StatusFacade = new StatusFacade();

  xit('should return initial State', () => {
    const state = statusReducer(TEST_INITIAL_STATE, noopAction);
    expect(state).toMatchSnapshot('initalState');
  });

  it('should work for loadNodes', () => {
    const action = facade.loadNodes();
    const state = statusReducer(TEST_INITIAL_STATE, action);
    expect(state).toMatchSnapshot('loadNodes');
  });

  it('should work for loadNodeStatus', () => {
    const action = facade.loadNodeStatus({ nodeId: TEST_NODE_ID, url: TEST_URL });
    const state = statusReducer(TEST_INITIAL_STATE, action);
    expect(state).toMatchSnapshot('loadNodeStatus');
  });

  it('should work for loadNodeStatusSuccess', () => {
    const action = facade.loadNodeStatusSuccess({ status: TEST_STATUS, nodeId: TEST_NODE_ID });
    const state = statusReducer(TEST_INITIAL_STATE, action);
    expect(state).toMatchSnapshot('loadNodeStatusSuccess');
  });

  it('should work for loadNodeStatusFailure', () => {
    const action = facade.loadNodeStatusFailure({ nodeId: TEST_NODE_ID });
    const state = statusReducer(TEST_INITIAL_STATE, action);
    expect(state).toMatchSnapshot('loadNodeStatusFailure');
  });

  it('should work for toggleNode', () => {
    const action = facade.toggleNode({ nodeId: TEST_NODE_ID });
    const state = statusReducer(TEST_INITIAL_STATE, action);
    expect(state).toMatchSnapshot('toggleNode');
  });
});
