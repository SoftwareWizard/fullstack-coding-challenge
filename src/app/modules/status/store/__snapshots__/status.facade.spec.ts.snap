// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer should work for loadNodeStatus: loadNodeStatus 1`] = `
Object {
  "nodes": Array [
    Object {
      "id": 1,
      "isExpanded": false,
      "isLoading": false,
      "isOnline": undefined,
      "name": "Node 1",
      "status": undefined,
      "url": "http://testurl.com",
    },
    Object {
      "id": 2,
      "isExpanded": false,
      "isLoading": true,
      "isOnline": undefined,
      "name": "Node 2",
      "status": undefined,
      "url": "http://testurl.com",
    },
  ],
}
`;

exports[`Reducer should work for loadNodeStatusFailure: loadNodeStatusFailure 1`] = `
Object {
  "nodes": Array [
    Object {
      "id": 1,
      "isExpanded": false,
      "isLoading": false,
      "isOnline": undefined,
      "name": "Node 1",
      "status": undefined,
      "url": "http://testurl.com",
    },
    Object {
      "id": 2,
      "isExpanded": false,
      "isLoading": false,
      "isOnline": false,
      "name": "Node 2",
      "status": undefined,
      "url": "http://testurl.com",
    },
  ],
}
`;

exports[`Reducer should work for loadNodeStatusSuccess: loadNodeStatusSuccess 1`] = `
Object {
  "nodes": Array [
    Object {
      "id": 1,
      "isExpanded": false,
      "isLoading": false,
      "isOnline": undefined,
      "name": "Node 1",
      "status": undefined,
      "url": "http://testurl.com",
    },
    Object {
      "id": 2,
      "isExpanded": false,
      "isLoading": false,
      "isOnline": true,
      "name": "testNodeName",
      "status": undefined,
      "url": "http://testurl.com",
    },
  ],
}
`;

exports[`Reducer should work for loadNodes: loadNodes 1`] = `
Object {
  "nodes": Array [
    Object {
      "id": 1,
      "isExpanded": false,
      "isLoading": false,
      "isOnline": undefined,
      "name": "Node 1",
      "status": undefined,
      "url": "http://testurl.com",
    },
    Object {
      "id": 2,
      "isExpanded": false,
      "isLoading": false,
      "isOnline": undefined,
      "name": "Node 2",
      "status": undefined,
      "url": "http://testurl.com",
    },
  ],
}
`;

exports[`Reducer should work for toggleNode: toggleNode 1`] = `
Object {
  "nodes": Array [
    Object {
      "id": 1,
      "isExpanded": false,
      "isLoading": false,
      "isOnline": undefined,
      "name": "Node 1",
      "status": undefined,
      "url": "http://testurl.com",
    },
    Object {
      "id": 2,
      "isExpanded": true,
      "isLoading": false,
      "isOnline": undefined,
      "name": "Node 2",
      "status": undefined,
      "url": "http://testurl.com",
    },
  ],
}
`;
